synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+        let timeout: Option<std::time::Instant> = None;
+
+        let futex = &(self.0);
+
+        loop {
+            // attempt to do the fast path
+            if futex.compare_and_swap(0, tid, Ordering::SeqCst) != 0 {
+                // we failed to do the fast path, wait in the kernel
+                let result = futex.lock_pi(1, None);
+
+                match result {
+                    // if we succeed, bail
+                    Ok(_) => break,
+
+                    // if we fail, we handle this a few different ways
+                    Err(NixError::Sys(errno)) => {
+                        if likely(errno == Errno::EINTR) {
+                            // EINTR means a signal interrupted our syscall
+                            // conditionally return or keep trying
+                            if signal_failure {
+                                return MutexLockResult::SignalFailure;
+                            } else {
+                                continue;
+                            }
+                        } else if timeout.is_some() && errno == Errno::ETIMEDOUT {
+                            // if we set a timeout and we timeout, explain that
+                            return MutexLockResult::Timeout;
+                        } else if errno == Errno::EDEADLK {
+                            // if this task already had the lock, give up
+                            panic!(
+                                "Somehow attempted to lock {:p} multiple times (blame {})",
+                                &self, tid
+                            );
+                        }
+
+                        // if we have no idea why we failed, tell the user that
+                        panic!("FUTEX_LOCK_PI with (futex={:p}({}), value=1, timeout=None) failed due to {}.",
+                            futex, futex.load(Ordering::SeqCst), errno);
+                    }
+                    // I hope to god this last one is impossible
+                    _ => {
+                        unreachable!("Unexpected error from futex wait (this should never happen)")
+                    }
+                }
+            } else {
+                // yay we don't have to use the kernel
+                break;
+            }
+        }
+
+        MutexLockResult::Ok
+    }
+
+    /// Check if this mutex is currently locked using only the fastpath.
+    fn islocked(&self) -> bool {
+        return self.0.load(Ordering::Relaxed) & FUTEX_TID_MASK == gettid();
+    }
+}
+
+unsafe impl<T: RawFutex> RawMutex for FutexMutex<T> {
+    const INIT: FutexMutex<T> = FutexMutex(T::INIT);
+
+    type GuardMarker = lock_api::GuardSend;
+
+    fn lock(&self) {
+        let result = self.mutex_get(false);
+        match result {
+            MutexLockResult::Ok => {}
+            _ => {
+                panic!("Failed to grab mutex, this should be impossible since signal_failure is false and no timeout.");
```

Comment:
Clarifying q -- why is this the case? Isn't it possible that the mutex init call above fails, since that doesn't seem to be handled over there?

Reply:
`mutex_get(false)` will either hang forever, successfully grab the mutex, or panic.

What I could probably do is refactor this such that mutex_get doesn't return a result, but this is nice because when we want mutexes to fail to lock if `EINTR` we have the same code.

bshrug